### PR TITLE
SI-8185 Correct grammar for single-warning compilation run

### PR DESCRIPTION
**Pull request by huitseeker**
50-odd updated tests (!)

review by @dragos or @adriaanm or anybody who isn't too busy (this is an easy one)

### CODE REVIEW

**Comment by huitseeker** [comment]
Phrasing chosen to be "1 xxx warning" rather than "a xxx warning" because this also covers inliner warnings. I didn't want to get into (a/an), being grammatically ok-ish is enough.

File: src/compiler/scala/tools/nsc/Global.scala [hunk 1]
```diff
@@ -1164,7 +1164,9 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
       else if (!(warnings contains pos)) warnings += ((pos, msg))
     def summarize() =
       if (warnings.nonEmpty && (option.isDefault || settings.fatalWarnings))
-        warning("there were %d %s warning(s); re-run with %s for details".format(warnings.size, what, option.name))
+        if (warnings.size > 1)
+          warning("there were %d %s warnings; re-run with %s for details".format(warnings.size, what, option.name))
+        else warning("there was one %s warning; re-run with %s for details".format(what, option.name))
```
Comment (soc) [inline src/compiler/scala/tools/nsc/Global.scala]: Why not use the f-interpolator here?

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/Global.scala]: Or the s-interpolator.

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/Global.scala]: Maybe you can tell I spent some time with the f-troop recently.

```
scala> val size = 5
size: Int = 5

scala> val what = "strange"
what: String = strange

scala> val opt = "demystify"
opt: String = demystify

scala> val x = new Formattable {
     | def formatTo(f: Formatter,g:Int,w:Int,p:Int) =
     | { val s = if (size>1) s"there were $size $what warnings"
     |   else s"there was one $what warning"
     |   f format s }}
x: java.util.Formattable = $anon$1@5e47f5ae

scala> f"$x; re-run with $opt"
res1: String = there were 5 strange warnings; re-run with demystify
```

**Comment by adriaanm** [comment]
Thanks, @huitseeker, that looks much nicer.

**Comment by adriaanm** [comment]
@scala/team-scala heads up: this PR potentially breaks check files that are in flight in the PR queue, so if we start seeing spurious failures, please rebase on top of this one. It may also cause test failures when merging 2.10.x into master.

**Comment by adriaanm** [comment]
[deferring my LGTM], because we may need to update checkfiles of PRs currently in flight. I'll take take care of this after flushing the queue.

**Comment by adriaanm** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
I'll look into updating the check files once the other PRs are in. PRs like this are fragile unfortunately due to or testing strategy

**Comment by huitseeker** [comment]
I can of course squash the two commits if you ping me. But with the intend-to-fix comments I wanted to leave @adriaanm with access to the old commit, which may be more composable w/ work he might have done already.

**Comment by adriaanm** [comment]
Thanks for the update. I know you submitted this PR a long time ago, but with our PR queue being so overloaded, we couldn't get to it in time for RC1. I'd still like to generate nicer warnings, but I don't think that qualifies as an RC blocker. I've set the milestone to 2.11.1-RC1, so we'll merge as soon as 2.11.0 goes final. As you've experienced first-hand, it's a time consuming undertaking to update all these check files, so it's certainly not a quick and low-cost fix for some low-hanging fruit.

**Comment by huitseeker** [comment]
Sure, no problem. I'll put a thread keeping this one up to date on the backburner.

**Comment by huitseeker** [comment]
Actually, @adriaanm, is there a way to pass `partest` options from ant ? That would help with maintaining this one. Just a `--show-diff` would be tremendously helpful.

**Comment by som-snytt** [comment]
@huitseeker aren't those already `diff *.log *.check` at the end of the build output?  If that's what you mean. I don't think you can pass arbitrary command line opts.

**Comment by huitseeker** [comment]
@som-snytt I'd have liked to swith from this `ant`-generated `diff` to a `partest`-generated `--show-diff` (more readable). And have the ability to do an `--update-check`, or `--verbose` ...

**Comment by huitseeker** [comment]
The failure on 26e6848 is expected and will disappear after squash. The one on 92d6ed9 is a network issue, unrelated to this PR.

**Comment by huitseeker** [comment]
PLS REBUILD/pr-scala@2d6ed9

**Comment by huitseeker** [comment]
PLS REBUILD/pr-scala@2d6ed9

(github pooped)

**Comment by huitseeker** [comment]
PLS SYNCH
NOLITTER!
PLS REBUILD/pr-scala@92d6ed9

**Comment by huitseeker** [comment]
PLS REBUILD/pr-scala@92d6ed9

**Comment by huitseeker** [comment]
Gaah ! I can't figure out what the kitteh does. And the build seems to consistentily trigger a fault in a "git clone" ?

I'll just squash and rebuild all.

**Comment by adriaanm** [comment]
Maybe because of the network outage at epfl yesterday?

**Comment by adriaanm** [comment]
It now went back to an actual check file difference. I suggest just running test/partest --update-check locally. We run the tests under -optimize during PR validation, but that shouldn't influence check files (if it does, the .flags file should include -optimize explicitly to document this and to simplify our workflow)

**Comment by adriaanm** [comment]
(You could of course also run `SCALAC_OPTS=-optimize test/partest --update-check`)

**Comment by huitseeker** [comment]
@adriaanm As for ["an actual check file difference"](https://github.com/scala/scala/pull/3419#issuecomment-37444160), it's not the failure you are looking for. Failure of the first commit in the unsquashed version of this PR was [expected](https://github.com/scala/scala/pull/3419#issuecomment-36044599). The [current state](https://github.com/huitseeker/scala/commit/33fa5f39f04eea09de886252f0245eeb56476e18) of the PR is nothing but the squash of the old one.

Labeling tests that depend on optimize is beyond the scope of this PR, alas. I'd suggest trying to run `SCALAC_OPTS=-optimize test/partest` to get a sense of the task.

NOLITTER!

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
Could you please resubmit this against the `2.11.x` branch? I am retiring the `master` branch today. Sorry about the long and circuitous path that this PR has taken, but I'm sure that correct grammar will prevail.

**Comment by som-snytt** [comment]
[As it must!](http://mediad.publicbroadcasting.net/p/wgcumain/files/201307/word_girl_core.jpg)

**Pull request closed without merging**